Refactor DocumentTimelineController to hold a set of AnimationTimelines

nmoucht opened the pull request:
#### bdef391506cd56d72c5d66467c5075b2b6ed01e7
<pre>
Refactor DocumentTimelineController to hold a set of AnimationTimelines
<a href="https://bugs.webkit.org/show_bug.cgi?id=279253">https://bugs.webkit.org/show_bug.cgi?id=279253</a>
<a href="https://rdar.apple.com/135402390">rdar://135402390</a>

Reviewed by Tim Nguyen.

In anticipation of implementing scroll-driven animations, refactor
DocumentTimelineController to be AnimationTimelineController, holding a set of
AnimationTimelines, so that ScrollTimelines and ViewTimelines can be updated
through the same mechanism as DocumentTimelines.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/DocumentTimeline.cpp:
(WebCore::DocumentTimeline::controller const):
(WebCore::DocumentTimeline::scheduleNextTick):
* Source/WebCore/animation/DocumentTimeline.h:
* Source/WebCore/animation/DocumentTimelinesController.cpp: Removed.
* Source/WebCore/animation/DocumentTimelinesController.h: Removed.
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::ScrollTimeline):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::ensureTimelinesController):
* Source/WebCore/dom/Document.h:
(WebCore::Document::timelinesController const):
* Source/WebCore/page/LocalFrame.cpp:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::scheduleRenderingUpdate):
(WebCore::Page::updateRendering):
(WebCore::Page::timelineControllerMaximumAnimationFrameRateDidChange):
(WebCore::Page::preferredRenderingUpdateFramesPerSecond const):
* Source/WebCore/page/Page.h:
* Source/WebCore/testing/Internals.cpp:

Canonical link: <a href="https://commits.webkit.org/283419@main">https://commits.webkit.org/283419@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/17526292de6a30be2c1e858774533b6f7835ad8e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/66082 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/45455 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18701 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/70114 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16692 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/68200 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/53254 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16973 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/53041 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11621 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/69149 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41936 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/57211 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33673 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38607 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/14588 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/15568 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/60501 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14935 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71816 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/10037 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/14349 "Found 1 new test failure: http/wpt/mediastream/transfer-mediastreamtrack-to-worker.html (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/60359 "Found 1 new test failure: imported/w3c/web-platform-tests/html/semantics/embedded-content/the-iframe-element/iframe_sandbox_anchor_download_block_downloads.tentative.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/10069 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/57273 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60650 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/8291 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1932 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10034 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/41263 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/42339 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/43522 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/42083 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -107,11 +112,11 @@ void DocumentTimelinesController::updateAnimationsAndSendEvents(ReducedResolutio
     Vector<Ref<CSSTransition>> completedTransitions;
     for (auto& timeline : protectedTimelines) {
-        auto shouldUpdateAnimationsAndSendEvents = timeline->documentWillUpdateAnimationsAndSendEvents();
+        auto shouldUpdateAnimationsAndSendEvents = is<DocumentTimeline>(timeline) ? dynamicDowncast<DocumentTimeline>(timeline)->documentWillUpdateAnimationsAndSendEvents() : DocumentTimeline::ShouldUpdateAnimationsAndSendEvents::No;
```
The fallback value here should be `Yes` actually, at least until we determine that other timeline types should have a similar function that determines whether they need an update (possibly if we determine that scroll offset hasn't changed since the last page update).

graouts replied:
Another thing we could do is to make that function not specific to `DocumentTimeline` and have the `AnimationTimeline` superclass just return `Yes` and let `DocumentTimeline` have its own override that does what it already does. That's probably a better thing to do. Can be done as a separate patch.

nmoucht replied:
Yeah I was planning on making a separate patch to add the `ScrollTimeline` and `ViewTimeline` objects to the `AnimationTimelinesController` and add the virtual functions on `AnimationTimeline`/implement the relevant ones on `ScrollTimeline` and `ViewTimeline`.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -52,32 +52,37 @@ DocumentTimelinesController::DocumentTimelinesController(Document& document)
-        timeline.suspendAnimations();
-    else
-        timeline.resumeAnimations();
+    if (is<DocumentTimeline>(timeline)) {
```
Maybe do an early `return` check here if we're _not_ dealing with a `DocumentTimeline`?

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -52,32 +52,37 @@ DocumentTimelinesController::DocumentTimelinesController(Document& document)
+    while (!m_timelines.isEmptyIgnoringNullReferences()) {
+        auto& timeline = *m_timelines.begin();
+        if (is<DocumentTimeline>(timeline))
+            dynamicDowncast<DocumentTimeline>(timeline)->detachFromDocument();
```
I believe this needs to happen for `ScrollTimeline` and `ViewTimeline` as well.

nmoucht replied:
This will be implemented in the patch I mentioned above.

nt1m commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -52,32 +52,38 @@ DocumentTimelinesController::DocumentTimelinesController(Document& document)
+    while (!m_timelines.isEmptyIgnoringNullReferences()) {
+        auto& timeline = *m_timelines.begin();
+        if (is<DocumentTimeline>(timeline))
+            dynamicDowncast<DocumentTimeline>(timeline)->detachFromDocument();
```
```suggestion
        if (RefPtr documentTimeline = dynamicDowncast<DocumentTimeline>(timeline))
            documentTimeline->detachFromDocument();
```

This is the more common pattern to avoid checking the pointer twice

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -107,11 +113,11 @@ void DocumentTimelinesController::updateAnimationsAndSendEvents(ReducedResolutio
     Vector<Ref<CSSTransition>> completedTransitions;
     for (auto& timeline : protectedTimelines) {
-        auto shouldUpdateAnimationsAndSendEvents = timeline->documentWillUpdateAnimationsAndSendEvents();
+        auto shouldUpdateAnimationsAndSendEvents = is<DocumentTimeline>(timeline) ? dynamicDowncast<DocumentTimeline>(timeline)->documentWillUpdateAnimationsAndSendEvents() : DocumentTimeline::ShouldUpdateAnimationsAndSendEvents::No;
```
Same here:

```suggestion
        RefPtr documentTimeline = dynamicDowncast<DocumentTimeline>(timeline);
        auto shouldUpdateAnimationsAndSendEvents = documentTimeline ? documentTimeline->documentWillUpdateAnimationsAndSendEvents() : DocumentTimeline::ShouldUpdateAnimationsAndSendEvents::No;
```

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -178,17 +184,21 @@ void DocumentTimelinesController::updateAnimationsAndSendEvents(ReducedResolutio
-        timeline->removeReplacedAnimations();
+    for (auto& timeline : protectedTimelines) {
+        if (is<DocumentTimeline>(timeline))
+            dynamicDowncast<DocumentTimeline>(timeline)->removeReplacedAnimations();
```
ditto

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -178,17 +184,21 @@ void DocumentTimelinesController::updateAnimationsAndSendEvents(ReducedResolutio
-        events.appendVector(timeline->prepareForPendingAnimationEventsDispatch());
+    for (auto& timeline : timelinesToUpdate) {
+        if (is<DocumentTimeline>(timeline))
+            events.appendVector(dynamicDowncast<DocumentTimeline>(timeline)->prepareForPendingAnimationEventsDispatch());
```
ditto

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -217,32 +227,36 @@ void DocumentTimelinesController::updateAnimationsAndSendEvents(ReducedResolutio
-        timeline->documentDidUpdateAnimationsAndSendEvents();
+    for (auto& timeline : timelinesToUpdate) {
+        if (is<DocumentTimeline>(timeline))
+            dynamicDowncast<DocumentTimeline>(timeline)->documentDidUpdateAnimationsAndSendEvents();
```
ditto

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -217,32 +227,36 @@ void DocumentTimelinesController::updateAnimationsAndSendEvents(ReducedResolutio
-        timeline.suspendAnimations();
+    for (auto& timeline : m_timelines) {
+        if (is<DocumentTimeline>(timeline))
+            dynamicDowncast<DocumentTimeline>(timeline)->suspendAnimations();
```
ditto

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -251,16 +265,18 @@ void DocumentTimelinesController::resumeAnimations()
-        timeline.resumeAnimations();
+    for (auto& timeline : m_timelines) {
+        if (is<DocumentTimeline>(timeline))
+            dynamicDowncast<DocumentTimeline>(timeline)->resumeAnimations();
```
ditto

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/Headers.cmake:
```diff
@@ -693,10 +693,11 @@ set(WebCore_PRIVATE_FRAMEWORK_HEADERS
     animation/CustomAnimationOptions.h
     animation/CompositeOperation.h
-    animation/DocumentTimelinesController.h
+
```
nit: remove blank line

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -52,32 +52,38 @@ DocumentTimelinesController::DocumentTimelinesController(Document& document)
+        dynamicDowncast<DocumentTimeline>(timeline)->suspendAnimations();
     else
-        timeline.resumeAnimations();
+        dynamicDowncast<DocumentTimeline>(timeline)->resumeAnimations();
```
```
    RefPtr documentTimeline = dynamicDowncast<DocumentTimeline>(timeline);
    if (!documentTimeline)
        return;

    if (m_isSuspended)
        documentTimeline->suspendAnimations();
    else
        documentTimeline->resumeAnimations();
```

nmoucht replied:
Fixed.

nt1m commented on Source/WebCore/animation/AnimationTimelinesController.cpp:
```diff
@@ -52,32 +52,38 @@ DocumentTimelinesController::DocumentTimelinesController(Document& document)
-        m_timelines.begin()->detachFromDocument();
+    while (!m_timelines.isEmptyIgnoringNullReferences()) {
+        auto& timeline = *m_timelines.begin();
+        if (RefPtr documentTimeline = dynamicDowncast<DocumentTimeline>(timeline))
+            documentTimeline->detachFromDocument();
```
Add FIXME to do that for other timelines?

nmoucht replied:
Added.

nt1m reviewed (approved):
I assume this looks good from @graouts' side too :)

nmoucht committed:
Refactor DocumentTimelineController to hold a set of AnimationTimelines
https://bugs.webkit.org/show_bug.cgi?id=279253
rdar://135402390

Reviewed by Tim Nguyen.

In anticipation of implementing scroll-driven animations, refactor
DocumentTimelineController to be AnimationTimelineController, holding a set of
AnimationTimelines, so that ScrollTimelines and ViewTimelines can be updated
through the same mechanism as DocumentTimelines.

* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/DocumentTimeline.cpp:
(WebCore::DocumentTimeline::controller const):
(WebCore::DocumentTimeline::scheduleNextTick):
* Source/WebCore/animation/DocumentTimeline.h:
* Source/WebCore/animation/DocumentTimelinesController.cpp: Removed.
* Source/WebCore/animation/DocumentTimelinesController.h: Removed.
* Source/WebCore/animation/ScrollTimeline.cpp:
(WebCore::ScrollTimeline::ScrollTimeline):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::ensureTimelinesController):
* Source/WebCore/dom/Document.h:
(WebCore::Document::timelinesController const):
* Source/WebCore/page/LocalFrame.cpp:
* Source/WebCore/page/Page.cpp:
(WebCore::Page::scheduleRenderingUpdate):
(WebCore::Page::updateRendering):
(WebCore::Page::timelineControllerMaximumAnimationFrameRateDidChange):
(WebCore::Page::preferredRenderingUpdateFramesPerSecond const):
* Source/WebCore/page/Page.h:
* Source/WebCore/testing/Internals.cpp:

Canonical link: https://commits.webkit.org/283419@main
